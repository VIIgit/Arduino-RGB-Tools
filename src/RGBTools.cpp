/**
 * @class RGBTools
 *
 * @brief Use this code with RGB-LEDs
 *
 * @detail This class implements methods to control a RGB LED with Arduino
 */

#include "RGBTools.h"

/**
 * constructor; saves the pins
 *
 * @param[in] r The Arduino pin number of the red pin of the LED
 * @param[in] g The Arduino pin number of the green pin of the LED
 * @param[in] b The Arduino pin number of the blue pin of the LED
 */
RGBTools::RGBTools(uint8_t r, uint8_t g, uint8_t b) {
    this->r_pin = r;
    this->g_pin = g;
    this->b_pin = b;
}

/**
 * Constructor using the Mode enum in header file.
 * Ex: RGBTools(9, 10, 11, COMMON_ANODE);
 *
 * @param[in] r The Arduino pin number of the red pin of the LED
 * @param[in] g The Arduino pin number of the green pin of the LED
 * @param[in] b The Arduino pin number of the blue pin of the LED
 * @param[in] mode The mode of the RGB (common cathode / common anode)
 */
RGBTools::RGBTools(uint8_t r, uint8_t g, uint8_t b, Mode mode) {
    this->r_pin = r;
    this->g_pin = g;
    this->b_pin = b;

    this->mode = mode;
}

/**
 * Set LED-color to custom color instantly
 *
 * @param[in] r The red value ([0-255])
 * @param[in] g The green value ([0-255])
 * @param[in] b The blue value ([0-255])
 */
void RGBTools::setColor(uint8_t r, uint8_t g, uint8_t b) {
    // set color of LED
    if (mode == COMMON_CATHODE) {
        analogWrite(r_pin, r);
        analogWrite(g_pin, g);
        analogWrite(b_pin, b);
    } else {
        analogWrite(r_pin, 255 - r);
        analogWrite(g_pin, 255 - g);
        analogWrite(b_pin, 255 - b);
    }

    // save state
    this->curr_r = r;
    this->curr_g = g;
    this->curr_b = b;
}

/**
 * Set LED-color to custom color instantly Uses a 32 bit value to set a color.
 *
 * @param[in] color The 32 bit color value (eg. 0xFF0000 for red)
 */
void RGBTools::setColor(uint32_t color) {
    uint8_t red = (uint8_t) ((color & 0xFF0000) >> 16);
    uint8_t green = (uint8_t) (color & 0x00FF00) >> 8;
    uint8_t blue = (uint8_t) (color & 0x0000FF);
    setColor(red, green, blue);
}

/**
 * Fades to custom color in specific time in specific steps
 *
 * @param[in] r The red value ([0-255])
 * @param[in] g The green value ([0-255])
 * @param[in] b The blue value ([0-255])
 * @param[in] steps The number of steps to use
 * @param[in] duration The duration of the fading in ms
 */
void RGBTools::fadeTo(uint8_t r, uint8_t g, uint8_t b, int steps, int duration) {
    // calculate difference to target
    uint8_t diff_r = r-curr_r;
    uint8_t diff_g = g-curr_g;
    uint8_t diff_b = b-curr_b;

    // calculate the width of each step multiplied by 100 as to not lose decimal data
    int steps_r = (diff_r *100 / steps);
    int steps_g = (diff_g *100 / steps);
    int steps_b = (diff_b *100 / steps);

    // calculate delay between steps
    int delay_value = (duration / steps);

    // loop through the steps (i: distance to target)
    for (int i = steps - 1; i >= 0; i--){

        // set color of current step
        this->setColor(
            r - (i * steps_r)/100,	// red target minus distance to target steps (divided by 100 as to offset steps_r calculations)
            g - (i * steps_g)/100,	// green
            b - (i * steps_b)/100 	// blue
        );

        // delay until next step
        delay((unsigned long) delay_value);
    }
}
