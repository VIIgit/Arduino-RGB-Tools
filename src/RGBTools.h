/*
* RGBTools
* Version 1.5 July, 2016
* 
* Use this code with RGB-LEDs.
*/

#ifndef RGBTools_H
#define RGBTools_H

#ifdef DEBUGGING
	#define analogWrite(pin,val) (printf("set %d to %d\n", pin, val))
	#define delay(val) (printf("delay %lu\n", val))
	#include <stdio.h>
	#include <stdint.h>
#else
	#include <Arduino.h>
#endif


enum Mode { COMMON_ANODE, COMMON_CATHODE };

class RGBTools
{
	public:
		RGBTools(uint8_t r, uint8_t g, uint8_t b);
		RGBTools(uint8_t r, uint8_t g, uint8_t b, Mode mode);
		void setColor(uint8_t r, uint8_t g, uint8_t b);
		void setColor(uint32_t);
		void fadeTo(uint8_t r, uint8_t g, uint8_t b, int steps, int duration);
		
	private:
		// pins for colors
		uint8_t r_pin, g_pin, b_pin;

		// saves current state (color)
		uint8_t curr_r, curr_g, curr_b;

		// output mode for common cathode or anode RGB LEDs
		uint8_t mode;

};

#endif
