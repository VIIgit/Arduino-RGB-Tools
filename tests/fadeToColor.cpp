#include <stdint.h>
#include "../src/RGBTools.h"

RGBTools rgb(1,2,3);

int main() {
  printf("# 255, 0, 0 in 3 steps, 5 sec\n");
  rgb.fadeTo(255, 0, 0, 3, 5000);
}