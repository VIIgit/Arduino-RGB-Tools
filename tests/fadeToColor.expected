# 255, 0, 0 in 3 steps, 5 sec
set 1 to 170
set 2 to 255
set 3 to 255
delay 1666
set 1 to 85
set 2 to 255
set 3 to 255
delay 1666
set 1 to 0
set 2 to 255
set 3 to 255
delay 1666
